pandas/core/indexing.py

```diff
@@ -429,11 +430,23 @@ def _setitem_with_indexer(self, indexer, value):
                     # this preserves dtype of the value
                     new_values = Series([value])._values
                     if len(self.obj._values):
-                        try:
-                            new_values = np.concatenate([self.obj._values, new_values])
-                        except TypeError:
+                        if is_timedelta64_dtype(
+                            new_values
+                        ) and not is_timedelta64_dtype(self.obj):
+                            # GH#22717 np.concatenate incorrect casts
+                            #  timedelta64 to integer
                             as_obj = self.obj.astype(object)
-                            new_values = np.concatenate([as_obj, new_values])
+                            new_values = np.concatenate(
+                                [as_obj, np.array([value], dtype=object)]
+                            )
+                        else:
+                            try:
+                                new_values = np.concatenate(
+                                    [self.obj._values, new_values]
+                                )
+                            except TypeError:
+                                as_obj = self.obj.astype(object)
```

Comment:
Was this hitting the except clause before? Not super familiar with this code but seems like some duplication we could ideally clean up / refactor

Reply:
It was not raising in the relevant case, and the np.concatenate would cast the timedelta64 to int64.

> seems like some duplication we could ideally clean up / refactor

Yes, this method in particular is in dire need of a refactor.  ATM I'm focused on `Block._try_coerce_arg` and `Block._can_hold_element` (see other open PRs) and I think/hope once some inconsistencies in those are ironed out, some of the special-casing here may be unnecessary.